### ISSUE TITLE
Crash with Generic Function

### ISSUE THREAD

**Issue by philberty** [Rust-GCC/gccrs#368]
@flip1995 made an interesting test case for generics not quite sure why this is crashing yet.

```rust
struct S;

fn foo<S>(s: S) -> S {
    s
}

fn main() {
    let _s: S = foo(S);
}
```

```
rust1: internal compiler error: in TyVar, at rust/typecheck/rust-tyty.cc:38
0x29422a4 diagnostic_impl(rich_location*, diagnostic_metadata const*, int, char const*, __va_list_tag (*) [1], diagnostic_t)
        ???:0
0x29440be internal_error(char const*, ...)
        ???:0
0x294443d fancy_abort(char const*, int, char const*)
        ???:0
0x1085c72 Rust::TyTy::TyVar::TyVar(unsigned int)
        ???:0
0x108b74b Rust::TyTy::ParamType::resolve() const
        ???:0
0x1090441 Rust::TyTy::BaseRules::unify(Rust::TyTy::BaseType*)
        ???:0
0x10870eb Rust::TyTy::ADTType::unify(Rust::TyTy::BaseType*)
        ???:0
0x1073faa Rust::Resolver::TypeCheckStmt::visit(Rust::HIR::LetStmt&)
        ???:0
0xfb81b1 Rust::HIR::LetStmt::accept_vis(Rust::HIR::HIRVisitor&)
        ???:0
0x1073c9d Rust::Resolver::TypeCheckStmt::Resolve(Rust::HIR::Stmt*, bool)
        ???:0
0x106531e Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::BlockExpr&)::{lambda(Rust::HIR::Stmt*)#1}::operator()(Rust::HIR::Stmt*)
        ???:0
0x1066eb7 std::_Function_handler<bool (Rust::HIR::Stmt*), Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::BlockExpr&)::{lambda(Rust::HIR::Stmt*)#1}>::_M_invoke(std::_Any_data const&, Rust::HIR::Stmt*&&)
        ???:0
0xed1b2c std::function<bool (Rust::HIR::Stmt*)>::operator()(Rust::HIR::Stmt*) const
        ???:0
0xec9317 Rust::HIR::BlockExpr::iterate_stmts(std::function<bool (Rust::HIR::Stmt*)>)
        ???:0
0x1065586 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::BlockExpr&)
        ???:0
0xfb6e91 Rust::HIR::BlockExpr::accept_vis(Rust::HIR::HIRVisitor&)
        ???:0
0x106c32b Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr*, bool)
        ???:0
0x107451b Rust::Resolver::TypeCheckItem::visit(Rust::HIR::Function&)
        ???:0
0xfb76eb Rust::HIR::Function::accept_vis(Rust::HIR::HIRVisitor&)
        ???:0
0x1074267 Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item*)
        ???:0
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
[Inferior 1 (process 137189) exited with code 04]
```

**Comment by philberty** [comment]
this will be fixed with: #155 and #385

**Comment by philberty** [comment]
update stack trace:

```
rust1: internal compiler error: in TyVar, at rust/typecheck/rust-tyty.cc:41
0x10a3434 Rust::TyTy::TyVar::TyVar(unsigned int)
        ../../gccrs/gcc/rust/typecheck/rust-tyty.cc:41
0x10a9c3b Rust::TyTy::ParamType::resolve() const
        ../../gccrs/gcc/rust/typecheck/rust-tyty.cc:1391
0x10aed93 Rust::TyTy::BaseRules::unify(Rust::TyTy::BaseType*)
        ../../gccrs/gcc/rust/typecheck/rust-tyty-rules.h:68
0x10a532b Rust::TyTy::ADTType::unify(Rust::TyTy::BaseType*)
        ../../gccrs/gcc/rust/typecheck/rust-tyty.cc:420
0xfd7631 Rust::HIR::LetStmt::accept_vis(Rust::HIR::HIRVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir-full-test.cc:4660
0x107b76f operator()
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check.cc:97
0x107d333 _M_invoke
        /usr/include/c++/9/bits/std_function.h:285
0xeec265 Rust::HIR::BlockExpr::iterate_stmts(std::function<bool (Rust::HIR::Stmt*)>)
        ../../gccrs/gcc/rust/hir/tree/rust-hir-expr.h:2484
0x107b97a Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::BlockExpr&)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check.cc:96
0xfd64cb Rust::HIR::BlockExpr::accept_vis(Rust::HIR::HIRVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir-full-test.cc:4174
0x1085b47 Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr*, bool)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check-expr.h:45
0xfd6d1b Rust::HIR::Function::accept_vis(Rust::HIR::HIRVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir-full-test.cc:4414
0x107b67f Rust::Resolver::TypeResolution::Resolve(Rust::HIR::Crate&)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check.cc:48
0xedcd70 Rust::Session::parse_file(char const*)
        ../../gccrs/gcc/rust/rust-session-manager.cc:569
0xedc8fe Rust::Session::parse_files(int, char const**)
        ../../gccrs/gcc/rust/rust-session-manager.cc:459
0xe39264 grs_langhook_parse_file
        ../../gccrs/gcc/rust/rust-lang.cc:171

```
